h/security/predicates.py

```diff
@@ -221,6 +221,10 @@ def get_authenticated_users_membership():
 def group_member_edit(
     identity, context: EditGroupMembershipContext
 ):  # pylint:disable=too-many-return-statements,too-complex
+    assert (
+        context.new_roles is not None
+    ), "new_roles must be set before checking permissions"
```

Comment:
The view always has to set `context.new_roles` before calling `request.has_permission(some_permission, context)` (which ends up calling this `group_member_edit()` predicate). Adding this `assert` here to make sure that the security predicate won't run if `new_roles` isn't set, as that could potentially lead to permissions being granted when they shouldn't be.

Reply:
**DD: Group Owners, Admins and Moderators** contains a [section explaining how security permissions work in Pyramid and h](https://docs.google.com/document/d/1XzRqt0NBAQCpyuES2vAh7fnbmng_mgAaWJQxz040LRA/edit?tab=t.0#heading=h.sjpzbyqckku5).